### PR TITLE
Extend op attrs with commutattive annotation

### PR DESCRIPTION
**Pull request by mikepapadim**
This PR extends the op `attrs` with commutativity. 
This is required for the ongoing work to use the pattern language to express more general patterns.

@mbs-octoml  @mbrookhart @jroesch @electriclilies

### CODE REVIEW

File: include/tvm/relay/op_attr_types.h [hunk 1]
```diff
@@ -89,6 +89,11 @@ using TNonComputational = bool;
  */
 using TReshapeOp = bool;
 
+/*!
+ * \brief Mark the operator as commutative to allow expr simplifications.
+ */
+using CommutativeOp = bool;
```
Comment (mbs-octoml) [inline include/tvm/relay/op_attr_types.h:95]: nit: TCommutativeOp

File: src/relay/ir/dataflow_matcher.cc [hunk 2]
```diff
@@ -213,6 +213,14 @@ bool DFPatternMatcher::VisitDFPattern_(const CallPatternNode* op, const Expr& ex
     }
     return false;
   };
+  auto is_commutative_op = [&get_op_node](const CallPatternNode* op_node) {
```
Comment (mbs-octoml) [inline src/relay/ir/dataflow_matcher.cc:216]: I was expecting this to look at the TCommutativeOp attribute of op_node.

File: src/relay/op/op_common.h [hunk 3]
```diff
@@ -60,6 +60,7 @@ namespace relay {
       .add_type_rel("Identity", IdentityRel)                                   \
       .set_attr<TOpPattern>("TOpPattern", kElemWise)                           \
       .set_attr<TOpIsStateful>("TOpIsStateful", false)                         \
+      .set_attr<CommutativeOp>("CommutativeOp", false)                         \
```
Comment (mbs-octoml) [inline src/relay/op/op_common.h:63]: well, I guess so...

File: src/relay/op/op_common.h [hunk 4]
```diff
@@ -84,6 +85,7 @@ namespace relay {
       .add_type_rel("Broadcast", BroadcastRel)                                          \
       .set_attr<TOpPattern>("TOpPattern", kBroadcast)                                   \
       .set_attr<TOpIsStateful>("TOpIsStateful", false)                                  \
+      .set_attr<CommutativeOp>("CommutativeOp", isCommutativeOp(OpName))                \
```
Comment (mbs-octoml) [inline src/relay/op/op_common.h:88]: What if it's non-comm by default (so no attrib binding), and the registration of (for now) add & multiply explicitly set it to true.

**Review by mbs-octoml — CHANGES_REQUESTED** [review]
Suggestion to default the attrib to false and explicitly set to true at registration site for add & mul. Adding more if op == "add" etc logic seems to be going against the grain.

File: src/relay/op/op_common.h (hunk 3 above)
Comment (electriclilies) [reply src/relay/op/op_common.h:63]: Do we need to put the commutativity property on unary and ops with 3+ arity?
I guess the downside is that then you'd need to check that arity == 2 before looking at the CommutativeOp attribute, or that the CommutativeOp attr is actually on the op.

**Comment by areusch** [comment]
closing due to inactivity, feel free to reopen if you'd like to merge

**Pull request closed without merging**